tests/learning/test_fault_tolerance.py

```diff
@@ -0,0 +1,191 @@
+
+
+def test_vladimir_cannot_verify_interface_with_ursulas_signing_key(blockchain_ursulas):
+    his_target = list(blockchain_ursulas)[4]
+
+    # Vladimir has his own ether address; he hopes to publish it along with Ursula's details
+    # so that Alice (or whomever) pays him instead of Ursula, even though Ursula is providing the service.
+
+    # He finds a target and verifies that its interface is valid.
+    assert his_target.interface_is_valid()
+
+    # Now Vladimir imitates Ursula - copying her public keys and interface info, but inserting his ether address.
+    vladimir = Vladimir.from_target_ursula(his_target, claim_signing_key=True)
+
+    # Vladimir can substantiate the stamp using his own ether address...
+    vladimir.substantiate_stamp(password=INSECURE_DEVELOPMENT_PASSWORD)
+    vladimir.stamp_is_valid()
+
+    # Now, even though his public signing key matches Ursulas...
+    assert vladimir.stamp == his_target.stamp
+
+    # ...he is unable to pretend that his interface is valid
+    # because the interface validity check contains the canonical public address as part of its message.
+    with pytest.raises(vladimir.InvalidNode):
+        vladimir.interface_is_valid()
+
+    # Consequently, the metadata as a whole is also invalid.
+    with pytest.raises(vladimir.InvalidNode):
+        vladimir.validate_metadata()
+
+
+def test_vladimir_uses_his_own_signing_key(blockchain_alice, blockchain_ursulas):
+    """
+    Similar to the attack above, but this time Vladimir makes his own interface signature
+    using his own signing key, which he claims is Ursula's.
+    """
+    his_target = list(blockchain_ursulas)[4]
+
+    fraduluent_keys = CryptoPower(power_ups=Ursula._default_crypto_powerups)
+
+    vladimir = Vladimir.from_target_ursula(target_ursula=his_target)
+
+    message = vladimir._signable_interface_info_message()
+    signature = vladimir._crypto_power.power_ups(SigningPower).sign(vladimir.timestamp_bytes() + message)
+    vladimir._interface_signature_object = signature
+
+    vladimir.substantiate_stamp(password=INSECURE_DEVELOPMENT_PASSWORD)
+
+    # With this slightly more sophisticated attack, his metadata does appear valid.
+    vladimir.validate_metadata()
+
+    # However, the actual handshake proves him wrong.
+    with pytest.raises(vladimir.InvalidNode):
+        vladimir.verify_node(blockchain_alice.network_middleware)
+
+
+def test_emit_warning_upon_new_version(ursula_federated_test_config, caplog):
+    lonely_ursula_maker = partial(make_federated_ursulas,
+                                  ursula_config=ursula_federated_test_config,
+                                  quantity=2,
+                                  know_each_other=True)
+    learner = lonely_ursula_maker().pop()
+    teacher, new_node = lonely_ursula_maker()
+
+    new_node.TEACHER_VERSION = learner.LEARNER_VERSION + 1
+
+    learner._current_teacher_node = teacher
+
+    warnings = []
```

Comment:
```suggestion
    warnings = list()
```

Reply:
https://giphy.com/gifs/DeoY3iC6VLBHG/html5

Reply:
https://gph.is/XIBFzF

Reply:
<iframe src="https://giphy.com/embed/DeoY3iC6VLBHG" width="480" height="213" frameBorder="0" class="giphy-embed" allowFullScreen></iframe><p><a href="https://giphy.com/gifs/come-at-me-bro-rejected-basketball-DeoY3iC6VLBHG">via GIPHY</a></p>

Reply:
:cactus:

Reply:
https://www.google.com/url?sa=i&rct=j&q=&esrc=s&source=images&cd=&ved=2ahUKEwi79b7TwpHfAhWBTt8KHfbRB5wQjRx6BAgBEAU&url=https%3A%2F%2Ftenor.com%2Fview%2Fblock-gumby-basketball-gif-5549764&psig=AOvVaw1ksMlAsjDigOuYzbEwxR6Q&ust=1544402730783155

Reply:
![tenor](https://user-images.githubusercontent.com/311973/49692171-f3592b80-fb21-11e8-82b5-a6bbb7da135c.gif)

Reply:
![https://giphy.com/gifs/tPrHcw5cz9hp6/html5]

Reply:
![gumby](https://giphy.com/gifs/tPrHcw5cz9hp6/html5)